refactor: extract CCheckQueue's data handling into a separate container "Bag"

martinus opened the pull request:
`CCheckQueue` has stored its work items in a `queue`, but made no guarantee about the order of elements in that container. This PR extracts that data storage handling into a separate container class `Bag`. This is pure refactoring, the result should have a better separation of concerns, adds tests for the new container, and makes it now easier to separately test and optimize the container.

martinus committed:
refactor: Bag: an unordered container

This introduces a simple container where elements can be added and
removed. The order of element removal is not specified and might change
in future due to e.g. optimizations.

The logic for that container is purely copied from CCheckQueue's current
implementation of the queue.

martinus committed:
refactor: use Bag in checkqueue

This simplfies CCheckQueue's Loop by using the new Bag container.

jonatack reviewed (commented):
Approach ACK 6b0537c4a71fe774ea12c097d5c16dc2a8ebb0a2

Needs trivial rebase, an adjacent Makefile entry and include header.

martinus commented:
Thanks @jonatack , I've rebased 6b0537c ->  6a9c6ea. Fixed Makefile & include conflict, and added a `reserve()` in a unit test to fix a clang-tidy warning.

fjahr commented:
I didn't dive too deeply into this yet but it's a bit sad that we can not remove more code due to this, that would certainly make it a more interesting change. Do you already have other places in mind where we could use this?

You also mention the order of elements, but it's unclear why that is important. This hasn't changed right? Or is this just about the naming choice?

I can't really make sense of the CI failure. Could you do a rebase? Since this is a fairly old change it might be some hidden merge conflict.

achow101 commented:
This PR does not seem to have attracted much attention from reviewers. As such, it does not seem important enough right now to keep it sitting idle in the list of open PRs.

Closing due to lack of interest.
